kernel: Add support for stopping workqueues

Mattemagikern opened the pull request:
This patch adds support for stopping workqueues. This is useful for freeing resources from workqueues when subsystems/modules is deactivated or cleaning up the system between tests in ztest to reach a fully normalized state.

The patch adds a new function k_work_queue_stop() that releases the workqueues thread and stack when a workqueue is unwanted. k_work_queue_stop(...) should be viewed as a counterpart to k_work_queue_start(...).

This would allow to:
k_work_queue_start(...);
<do work>
k_work_drain(..., true);
k_work_queue_stop(...);

andyross commented on kernel/work.c:
```diff
@@ -812,6 +812,36 @@ int k_work_queue_unplug(struct k_work_q *queue)
+	}
+
+	k_thread_abort(&queue->thread);
+	k_thread_join(&queue->thread, K_FOREVER);
```
On a technical level: you don't need to do both of these.  k_thread_abort() is already synchronous and will only return when the thread is dead.  Join, at that point, is a guaranteed noop unless something else re-starts the thread from another context, which won't happen because you hold the queue's lock.

But you surely don't want to use abort at all, as it will destroy the target thread in whatever state it might be in, which can be in app code inside callbacks holding resources you can't predict or control.

IMHO the correct technique here is to set an "exit" flag and have the work queue's main loop exit the thread on its own, which you can then reap here with join().

andyross replied:
And obviously that flag and check could share a lot of implementation with drain()

Mattemagikern replied:
Yepp, agreed.
I'll make the appropriate changes ASAP. 

Thanks for pointing it out.

cfriedt commented:
LGTM, but I'll wait until others have reviewed too.

andyross commented on tests/kernel/workq/work_queue/src/start_stop.c:
```diff
@@ -0,0 +1,83 @@
+static void work_handler(struct k_work *work)
+{
+	atomic_inc(&num_work_executed);
+	k_free(work);
```
Can't do this: the k_work object is (I'm like 70% certain, and if not it absolutely should be expected to be unless we explicitly document that the pointer is now unused) still in use inside the queue code that called the handler and you're creating a free memory bug.

Really heap usage in Zephyr tests is discouraged (except obviously in heap tests, heh).  It's not the native idiom, and test code tends to rely heavily on static allocations.  That's not a firm rule, but in this case I don't see any particular value the malloc/free are bringing.

Mattemagikern replied:
I was worried about just this haha. 

I'll get it sorted to static memory!

andyross commented on tests/kernel/workq/work_queue/prj.conf:
```diff
@@ -2,6 +2,7 @@ CONFIG_ZTEST=y
 CONFIG_POLL=y
 
 CONFIG_ASSERT=y
+CONFIG_HEAP_MEM_POOL_SIZE=512
```
See comment elsewhere about heap usage

andyross reviewed (changes requested):
Feature in general looks clean and useful.

One bug in the test.  Some criticism of IMHO an inappropriately aggressive implementation of "stop()".

And just as general hygiene, can you split this up into three patches:

1. Add the feature
2. Add tracing hooks
3. Add tests

Mattemagikern commented:
> And just as general hygiene, can you split this up into three patches:
> 
> 1. Add the feature
> 2. Add tracing hooks
> 3. Add tests

I'll split them up if you want, but could I ask why you prefere it that way?
Just for my own curiosity.

cfriedt commented:
@Mattemagikern - separate commits make it easier to review and maintain in many cases. As general rules, to ensure if a code change needs to be backported, there are fewer side effects with smaller commits. If a regression occurs, it's easier to narrow down the root cause. We usually insist on separate code and test commits because it makes it easier to test bug fixes.

Generally good hygiene

Mattemagikern committed:
kernel: Add support for stopping workqueues

This patch adds support for stopping workqueues. This is useful for freeing
resources from workqueues when subsystems/modules is deactivated or
cleaning up the system between tests in ztest to reach a fully normalized
state.

The patch adds a new function k_work_queue_stop() that releases the
workqueues thread and stack when a workqueue is unwanted.
k_work_queue_stop(...) should be viewed as a counterpart to
k_work_queue_start(...).

This would allow to:
k_work_queue_start(...);

k_work_drain(..., true);
k_work_queue_stop(...);

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

andyross commented:
> I'll split them up if you want, but could I ask why you prefere it that way? Just for my own curiosity.

Once upon a time on Discord:

_"Help help my very important code stopped working!"_

_"Maybe it's the new feature that just got added.  Can you try reverting this patch and see if that fixes it?"_

_"Help help!  I tried git revert and now it's giving me a bunch of collisions in the tracing code!  Who can help me fix this?  My code is very important!"_

Small patches that do one thing in an isolated way makes it easier to figure out what broke when things inevitably break.  Maintainers are lazy and like to toss off one-line suggestions vs. addressing rebase failures in their spare time.

Mattemagikern commented:
@cfriedt @andyross Thanks for the explination! 

The patch has been updated to only contain the functionallity. 
I've confirmed the changes with my testcases seperately.

Mattemagikern committed:
tracing: kernel: Add support for stopping workqueues

Adds tracing support for stopping workqueues.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

Mattemagikern committed:
tests: kernel: Add support for stopping workqueues

This patch adds tests for stopping workqueues.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

Mattemagikern commented:
Another newbi question: 
When you say you want the patches in [feature, tracing, tests], are these separate pull-requests or should I just split it up in commits on this pull-request?

npitre commented:
Separate commits within the same pull request please.

Mattemagikern commented:
@npitre thanks! :heavy_check_mark:

andyross commented on include/zephyr/kernel.h:
```diff
@@ -3606,6 +3606,19 @@ int k_work_queue_drain(struct k_work_q *queue, bool plug);
+ * @retval -ENODEV if the work queue was not started
+ * @retval -EBUSY if the work queue is actively processing work items
+ */
+int k_work_queue_stop(struct k_work_q *queue);
```
Note that this is a blocking call (because of k_thread_join()) and so per "Standard Zephyr API Idiom" should probably take a k_timeout to allow the app to control wakeup behavior.  In this particular case that's probably not a big deal because the only way this will get stuck is if a workq handler refuses to return, which is already a bug elsewhere.

Mattemagikern replied:
I'm worried about side effects if an user sets a very short timeout, e.g. `K_NO_WAIT` and/or thread priorities. 
Where we get a false negative. 

The case I'm worried about is: 
```
int k_work_queue_stop(struct k_work_q *queue, k_timeout_t delay)
{
...
	flag_set(&queue->flags, K_WORK_QUEUE_STOP_BIT);
	notify_queue_locked(queue);
	k_spin_unlock(&lock, key);
	if (k_thread_join(&queue->thread, timeout)) { <--- sleeping here
		key = k_spin_lock(&lock);
		flag_clear(&queue->flags, K_WORK_QUEUE_STOP_BIT);
		k_spin_unlock(&lock, key);
		return -ETIMEOUT;
	}

	return 0;

...

static void work_queue_main(void *workq_ptr, void *p2, void *p3)
{
....
		k_spinlock_key_t key = k_spin_lock(&lock); <-- past this point
....		
}

		
k_work_queue_stop(queue, K_NO_WAIT); // returns -ETIMEOUT
// Is the thread on it's way to die or is it alive? 
```
It may still be ok? 
Subsequent calls to the workqueue will be handled (-ENODEV) due to the thread resets the flags.
I may be overthinking this. What do you think?

andyross replied:
If `k_work_queue_stop()` is guaranteed to set the `STOP` bit, then the work queue is guaranteed to stop in finite time (absent app bugs that stall the handlers).  I think that's acceptable.  Apps that care about synchronization can always call it again and take care to avoid racing with their own handlers until they get a success.

Again, I don't think this rises to a -1 given the specifics; the timeout isn't likely to be very useful in practice.  Nonetheless timeout arguments for blocking calls are The Zephyr Way, so it should be pointed out.

andyross commented on include/zephyr/kernel.h:
```diff
@@ -3606,6 +3606,19 @@ int k_work_queue_drain(struct k_work_q *queue, bool plug);
+/** @brief Stop a work queue.
+ *
+ * This stops the work queue thread and releases its resources. This should be
+ * invoked when the work queue is no longer needed.
```
"Releases its resources" is a little unspecific.  This stops a thread, but doesn't actually "release" anything as the thread object remains owned by the queue.

It also wouldn't hurt to explicitly point out that the call is synchronous: the thread is stopped before the call returns so the app is guaranteed that no further handlers will be in progress or called in the future.

andyross reviewed (dismissed):
A few more nitpicks that occur to me, but this looks good.

cfriedt commented on tests/kernel/workq/work_queue/src/start_stop.c:
```diff
@@ -0,0 +1,82 @@
+{
+	size_t i, j;
+
+	for (i = 0; i < NUM_TEST_ROUNDS; i++) {
```
It isn't clear that this for loop increases test coverage beyond executing things once.

There are likely other tests that test regular execution of a work queue, such as queueing multiple work queue items, so I would suggest narrowing the scope of this test to only test the function added in this PR - `k_work_queue_stop()`

cfriedt commented on tests/kernel/workq/work_queue/src/start_stop.c:
```diff
@@ -0,0 +1,82 @@
+
+static void workq_setup(void)
+{
+	zassert_equal(k_work_queue_stop(&work_q), -ENODEV,
```
`-ENODEV` (no such device??) seems like a very odd return code in this case.

Why not change this to `-EALREADY`?

cfriedt commented on tests/kernel/workq/work_queue/src/start_stop.c:
```diff
@@ -0,0 +1,82 @@
+	.name = "test_work_q",
+	.no_yield = true,
+};
+
```
The contents of the setup function here should probably just be inserted into the `ZTEST()` below

cfriedt commented on tests/kernel/workq/work_queue/src/start_stop.c:
```diff
@@ -0,0 +1,82 @@
+
+	for (i = 0; i < NUM_TEST_ROUNDS; i++) {
+		workq_setup();
+		for (j = 0; j < NUM_TEST_ITEMS; j++) {
```
Similarly, I don't think this loop needs to be here since it likely does not increase coverage.

cfriedt commented on tests/kernel/workq/work_queue/src/start_stop.c:
```diff
@@ -0,0 +1,82 @@
+#include <zephyr/kernel.h>
+#include <zephyr/ztest.h>
+
+#define NUM_TEST_ROUNDS 5
+#define NUM_TEST_ITEMS  10
```
We shouldn't need to execute either of these loops more than once.

Let's narrow the scope of what we are testing to the functionality being added and mainly test things relating to `k_work_queue_stop()`

cfriedt commented on tests/kernel/workq/work_queue/src/start_stop.c:
```diff
@@ -0,0 +1,82 @@
+static struct k_work_q work_q;
+static K_THREAD_STACK_DEFINE(work_q_stack, STACK_SIZE);
+
+struct k_work works[NUM_TEST_ITEMS];
```
```suggestion
static struct k_work works[NUM_TEST_ITEMS];
```

cfriedt commented on tests/kernel/workq/work_queue/src/start_stop.c:
```diff
@@ -0,0 +1,82 @@
+
+struct k_work works[NUM_TEST_ITEMS];
+
+static atomic_t num_work_executed;
+static void work_handler(struct k_work *work)
+{
+	ARG_UNUSED(work);
+
+	atomic_inc(&num_work_executed);
+}
```
It probably isn't necessary to use atomics here (or even have a counter).

I would suggest adding a simple 100ms sleep inside of the work handler.

That should be sufficient time to verify that the work queue is busy in the main test thread.

cfriedt commented on tests/kernel/workq/work_queue/src/start_stop.c:
```diff
@@ -0,0 +1,82 @@
+			   K_PRIO_PREEMPT(4), &cfg);
+}
+
+static void workq_teardown(void)
```
I think the contents of this function should be inserted into the `ZTEST()` below

cfriedt commented on tests/kernel/workq/work_queue/src/start_stop.c:
```diff
@@ -0,0 +1,82 @@
+			zassert_equal(k_work_submit_to_queue(&work_q, &works[j]), 1,
+				      "Failed to submit work item");
+		}
+		/* Wait for the work item to complete */
+		k_sleep(K_MSEC(CHECK_WAIT));
+
+		/* Check that the work item was executed */
+		zassert_equal(atomic_get(&num_work_executed), i * j + j,
+			      "all work items were not executed within the expected time");
```
```suggestion
```

cfriedt commented on tests/kernel/workq/work_queue/src/start_stop.c:
```diff
@@ -0,0 +1,82 @@
+	}
+}
+
+ZTEST_SUITE(workqueue_reuse, NULL, NULL, NULL, NULL, NULL);
```
Ideally, rather than adding a whole new testsuite effectively for a single ZTEST testing a single API call, this should be added to an existing testsuite that are more like traditional unit tests for the work queue API.

Mattemagikern replied:
in the build project there exists the: 

> workqueue_triggered
and 
> workqueue_delayed

They seem more specifc than my test, should I rename my suit to workqueue_api for future use instead?

cfriedt commented on tests/kernel/workq/work_queue/src/start_stop.c:
```diff
@@ -0,0 +1,82 @@
+		      "Succeeded to submit work item to non-initialized work queue");
+}
+
+ZTEST(workqueue_reuse, test_submit_to_queue)
```
Let's narrow the scope of the code we are testing to `k_work_queue_stop()` and add it to some other testsuite, since it is unlikely that this test is adding new coverage for work queue execution of handling multiple work items.

It doesn't really make sense to create a new suite for effectively testing one API call.

If we don't have an existing unit-style testsuite (e.g. `workqueue_api`), maybe this could be it.

```suggestion
ZTEST(workqueue_api, test_k_work_queue_stop)
```

Mattemagikern replied:
I'll wait with these changes until we've decided on the correct placement of the tests :+1:

cfriedt commented on include/zephyr/kernel.h:
```diff
@@ -3606,6 +3606,19 @@ int k_work_queue_drain(struct k_work_q *queue, bool plug);
+ * @param queue pointer to the queue structure.
+ *
+ * @retval 0 if the work queue was stopped
+ * @retval -ENODEV if the work queue was not started
```
```suggestion
 * @retval -EALREADY if the work queue was not started (or already stopped)
```

Mattemagikern replied:
Done!

TaiJuWu commented on kernel/work.c:
```diff
@@ -653,6 +653,12 @@ static void work_queue_main(void *workq_ptr, void *p2, void *p3)
 			 * submissions.
 			 */
 			(void)z_sched_wake_all(&queue->drainq, 1, NULL);
+		} else if (flag_test(&queue->flags, K_WORK_QUEUE_STOP_BIT)) {
+			/* User has requested that the queue stop. Clear the status flags and exit.
+			 */
+			flags_set(&queue->flags, 0);
+			k_spin_unlock(&lock, key);
+			return;
```
Pardon me.
If we allow task append to the queue after setting `K_WORK_QUEUE_STOP_BIT`, the block will never be executed.

Mattemagikern replied:
A pre-condition to the _stop(..) function is that the queue has already been plugged. 

```
static inline int queue_submit_locked(struct k_work_q *queue,
				      struct k_work *work)
{
	 * * -ENODEV if the queue isn't running.
	 * * -EBUSY if draining and not chained
	 * * -EBUSY if plugged and not draining
	 * * otherwise OK

```

So the case you're thinking of is: 

Thread a. Continously adds work to queue. 

Thread b. pluggs queue & tries to stop the queue with K_FOREVER. 

Would thread b. freeze indefenately? 
Initially, maybe :+1: 

Question is if we should disallow that append of work if the K_WORK_QUEU_STOP_BIT is set as well. 
I'm not sure. How do you reason?

TaiJuWu replied:
Oh, I missed we need to call `k_work_queue_drain` before stopping `workqueue`, but the design make me thinks about the scenario will happen?
```
Thread a {
          while (true) {
                   queue_submit_locked(....);
         }
}

Thread b    k_work_queue_drain(..., true);
Thread c    k_work_queue_drain(..., true);
Thread b    k_work_queue_stop(..., K_FOREVER);
Thread c    k_work_queue_unplug(...);
```

cfriedt replied:
Sounds like a great opportunity for additional test cases

Mattemagikern replied:
@TaiJuWu 

Another option is to forbid the unplug function if `K_WORK_QUEUE_STOP_BIT` is set. 
```
Thread a {
          while (true) {
                   queue_submit_locked(....); <- Error once K_WORK_QUEUE_PLUGGED_BIT is set
         }
}

Thread b    k_work_queue_drain(..., true);
Thread c    k_work_queue_drain(..., true);
Thread b    k_work_queue_stop(..., K_FOREVER); <- Would exit in finite time
Thread c    k_work_queue_unplug(...); <- Would yield error & not unplug. 
```
Either thread B or C would return an error, depending on which accessed the queue first.
I believe this is the smallest change we could make to implement this feature. It shouldn't affect current use cases, what do you think? 

@cfriedt Agreed!

Mattemagikern replied:
The case I'm making could still deadlock thread b, if the queue never drains the queue will continue to accept works. 

So we probbably need the 2. You outlined earlier as well.
> 2. Disallow append if `K_WORK_QUEUE_STOP_BIT` is set.

But that changes the behaviour of the queue & drain system slightly, is that ok?

TaiJuWu replied:
> Another option is to forbid the unplug function if K_WORK_QUEUE_STOP_BIT is set.

Sound great.

>But that changes the behaviour of the queue & drain system slightly, is that ok?

It is ok for me but we still need @andyross 's or other maintainers comfirm.

Mattemagikern replied:
> By the way, I am not sure we need to call k_work_queue_drain before k_work_queue_stop by users is good api design or not.

Myself, being biased (haha), think it is neat :P The `queue_drain(..., true) ` gives the user the knowledge that the works scheduled up to that point will be executed and in so doing letting free-routines kick in as to not casue memory leaks in the application. 
`The k_work_queue_stop(...)` stopps the thread as soon as the last job has been processed. 

I'll commit the changes that we outlined here today :)

TaiJuWu replied:
I think we need to call `queue_drain(..., true)` in `k_work_queue_stop` and set `K_WORK_QUEUE_STOP_BIT ` berfore `queue_drain(..., true)`
Is it possible?
```
TA                                     TB
queue_drain(..., true)
                                    k_work_queue_stop but flag is not set
queue_unplug
                                   flag is set

```

Mattemagikern replied:
> Additionally, [System Workqueue](https://docs.zephyrproject.org/latest/kernel/services/threads/workqueue.html#id4)[](https://docs.zephyrproject.org/latest/kernel/services/threads/workqueue.html#system-workqueue) should never stop I guess. 

The patch specifically addresses scenarios where stopping a workqueue becomes interesting, like in a couple of tests I have in mind. 
I believe the patch aligns with the broader system goals. If this approach conflicts with project guidelines or expectations, I expect that the maintainers wil let me know shortly haha :smiling_face_with_tear:  

> I think we need to call queue_drain(..., true) in k_work_queue_stop internally and set K_WORK_QUEUE_STOP_BIT  berfore queue_drain(..., true)
Is it possible in current design? 

I've just looked through the `kernel/work.c` file, it would be a neat addition, one less function call from the users to achive the same effect. What's stopping me is the `k_work_queue_drain(...)` since the current implementation is blocking ( `K_FOREVER` ) and to resolve that I would need to rework that function to include a timeout and then manage to make the timeout time be split between the draining & the stopping of the thread (the stopping of the queue would be, almost, instantanious at that point if we've passed the drain stage). 
@cfriedt @andyross How would you like me to proceede with: 

-  to sort out the`_drain(..)` to include a timeout and incorperate parts of that function in the `_stop(...)` 

or

- keep it as is with the addition of blocking `_unplug(..)` and stop new works to be scheduled if the stop bit is set? 

Regarding the race-condition you're (@TaiJuWu) outlining I don't think it is a problem since the getting & setting of the flags are protected by the spin_lock, the winner of the spin_lock would dictate the outcome.

andyross commented:
FWIW:

> If we allow task append to the queue continuously after setting K_WORK_QUEUE_STOP_BIT, the block will never be executed.

I think that's more of a policy decision than a bug.  I can see arguments either way: you could force the queue to reject new submissions when a STOP is pending, or you could allow it and document it as "the queue will stop processing the next time it is idle".  Either one seems like a reasonable choice.

andyross reviewed (approved):
Forgot to refresh +1.  I'm fine with this as-is.  Most of the suggestions seem acceptable too.

cfriedt commented on tests/kernel/workq/work_queue/src/start_stop.c:
Probably would be best to rename to `workq/worq_api`, but I'm not blocking.

TaiJuWu reviewed (approved):
I don’t want to block this feature too.
